src/sysctl/sysctl.c

```diff
@@ -83,24 +154,29 @@ static int apply_all(OrderedHashmap *sysctl_options) {
         ORDERED_HASHMAP_FOREACH(option, sysctl_options, i) {
                 int k;
 
-                k = sysctl_write(option->key, option->value);
-                if (k < 0) {
-                        /* If the sysctl is not available in the kernel or we are running with reduced
-                         * privileges and cannot write it, then log about the issue, and proceed without
-                         * failing. (EROFS is treated as a permission problem here, since that's how
-                         * container managers usually protected their sysctls.) In all other cases log an
-                         * error and make the tool fail. */
-
-                        if (option->ignore_failure || k == -EROFS || ERRNO_IS_PRIVILEGE(k))
-                                log_debug_errno(k, "Couldn't write '%s' to '%s', ignoring: %m", option->value, option->key);
-                        else if (k == -ENOENT)
-                                log_info_errno(k, "Couldn't write '%s' to '%s', ignoring: %m", option->value, option->key);
-                        else {
-                                log_error_errno(k, "Couldn't write '%s' to '%s': %m", option->value, option->key);
-                                if (r == 0)
-                                        r = k;
-                        }
-                }
+                k = sysctl_write_or_warn(option->key, option->value, option->ignore_failure);
+                if (r == 0)
+                        r = k;
+
+                /* "Propagate" net.*.conf.default settings to interfaces which were already created before we
+                 * managed to write the default value. This is done to close the race between network
+                 * interface probing and systemd-sysctl.service running at boot.
+                 *
+                 * The keys use slashes as separators and are normalized, see sysctl_normalize above. */
+                _cleanup_free_ char *prefix = NULL, *rest = NULL;
+                int n;
+
+                k = sscanf(option->key, "net/%*[^/]/conf%n/default/%ms", &n, &rest);
+                if (k != 1)
+                        continue;
+
+                prefix = strndup(option->key, n);
+                if (!prefix)
+                        return log_oom();
+
+                k = copy_net_default(sysctl_options, prefix, rest, option->value, option->ignore_failure);
+                if (r == 0)
+                        r = k;
```

Comment:
uh, oh, are you sure this actually works? iirc the inheritance algorithm for each individual sysctl is actually very unique and different from all others, hence such blanket copying sounds really dangerous to me. I doubt this can work correctly? Have you verified that this is an OK thing to do?

It was my understanding that the handling of the network sysctls in the kernel is just generally fucked an awkward and basically unfixable from userspace. i.e. the "default", "all" and per-interface sysctls interact in amazing strange ways, and different for each option.

Reply:
I didn't do any deep research, but my understanding is that "default" when set applies to any interfaces which are problem later. If we assume that the user does not control the ordering between systemd-sysctl running and interfaces being probed, then the user cannot know if "default" will apply to some given interface or not, and this patch just makes the default apply to all interfaces predictably.

That said, the assumption in previous para is not universally valid. In particular, the user might know that a given interface is present in the initramfs, so any systemd-sysctl running in the real root must always be later, so if the default is only configured in real root, it will not apply to that interface. But I think we can accept that change in behaviour.

The new arrangement still seems generally better: "default" applies everywhere by default unless explicitly overridden, and individual interfaces can be configured differently.